src/mbed_cloud/connect/notifications.py

```diff
@@ -200,26 +212,151 @@ def __init__(self, db, queues, b64decode=True, notifications_api=None,
 
         self._stopping = False
         self._stopped = threading.Event()
+        self._ws = None
+        self._api_key = notifications_api.api_client.configuration.api_key['Authorization']
+        self._host = notifications_api.api_client.configuration.host
+        self._force_clear = force_clear
+        self._logger = logger
+        self._api_client = notifications_api.api_client
 
     @catch_exceptions(mds.rest.ApiException)
     @functools.wraps(threading.Thread.run)
     def run(self):
         """Thread main loop"""
         try:
-            while not self._stopping:
-                data = self.notifications_api.long_poll_notifications()
-                handle_channel_message(
-                    db=self.db,
-                    queues=self.queues,
-                    b64decode=self._b64decode,
-                    notification_object=data
-                )
-                if self.subscription_manager:
-                    self.subscription_manager.notify(data.to_dict())
+            self._start_notifications()
         finally:
             self._stopped.set()
 
+    def _get_on_message_calback(self):
+        def on_message(ws, data):
+            if self._stopping:
+                self.stop()
+            if self._logger:
+                self._logger.debug('received notification data: %s', data)
+            try:
+                data = self._api_client.deserialize(NotificationsThread.NotificationWebsocketMessage(data),
+                                                    NotificationMessage)
+            except Exception as e:
+                if self._logger:
+                    self._logger.error('Could not deserialise received notification: %s because %s ', data, e)
+
+            handle_channel_message(
+                db=self.db,
+                queues=self.queues,
+                b64decode=self._b64decode,
+                notification_object=data
+            )
+            if self.subscription_manager:
+                self.subscription_manager.notify(data.to_dict())
+
+        return on_message
+
+    def _get_on_close_callback(self):
+        def on_close(ws, error, reason):
+            if error == 1000:
+                self._delete_websocket_channel()
+            elif error == 1008:
+                self._log_error(error, reason)
+            elif error == 1006:
+                self._start_websocket()
```

Comment:
I think calling the start websocket method which calls the run_forever is eventually going to blow the stack as it is recursive and will recurse every time the socket closes.